src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs

```diff
@@ -37,10 +44,21 @@ class PatternVariableFinder : CSharpSyntaxWalker
 
             var result = finder.declarationPatterns;
             finder.declarationPatterns = null;
+            Debug.Assert(finder.expressionsToVisit.Count == 0);
             s_poolInstance.Free(finder);
             return result;
         }
 
+        private void VisitExpressions()
+        {
+            while (expressionsToVisit.Count != 0)
```

Comment:
This is doesn't preserve the order in which expressions are visited, which leads to a wrong behavior when duplicate duplicate declarations are involved. Please fix.